src/clj_irods/core.clj

```diff
@@ -0,0 +1,205 @@
+            single-item (icat/cached-get-item irods user zone path)
+            listing-item (when (delay? paged-folder-listing) (get-from-listing (apply concat (icat/flatten-cached-listings @paged-folder-listing))))]
+        (or
+          (when listing-item (delay (extract-fn listing-item)))
+          (when single-item  (delay (extract-fn @single-item)))
+          nil))
+      (or
+        ;; better if we pass user group IDs in so we can get them from any source, rather than forcing the icat version specifically
+        (when (:has-icat irods) (delay (force (icat/user-group-ids irods user zone)) (extract-fn @(icat/get-item irods user zone path))))
+        nil))))
+
+(defn- from-stat
+  [cache? irods extract-fn path]
+  (if cache?
+    (let [jargon-stat (jargon/cached-stat irods path)]
+      (or
+        (when jargon-stat (delay (extract-fn @jargon-stat)))
+        nil))
+    (or
+      (when (:has-jargon irods) (delay (extract-fn @(jargon/stat irods path))))
+      nil)))
+
+(defn- from-stat-or-listing
+  [stat-extract-fn listing-extract-fn irods user zone path]
+  (cached-or-get irods
+    [from-listing listing-extract-fn user zone path]
+    [from-stat stat-extract-fn path]))
+
+(defn object-type
+  [irods user zone path]
+  (let [listing-extract-fn (fn [item] (condp = (:type item) "dataobject" :file "collection" :dir nil))]
+    (from-stat-or-listing :type listing-extract-fn irods user zone path)))
+
+(defn date-modified
+  [irods user zone path]
+  (from-stat-or-listing :date-modified (comp (partial * 1000) #(Integer/parseInt %) :modify_ts) irods user zone path))
+
+(defn date-created
+  [irods user zone path]
+  (from-stat-or-listing :date-created (comp (partial * 1000) #(Integer/parseInt %) :create_ts) irods user zone path))
+
+(defn file-size
+  [irods user zone path]
+  (from-stat-or-listing :file-size :data_size irods user zone path))
+
+(defn uuid
+  [irods user zone path]
+  (cached-or-get irods
+    [from-listing :uuid user zone path]
+    [(fn [cache? irods path]
+       (let [get-from-metadata (fn [metadata]
+                                 (when-let [uuid-meta (first (filter #(= (:attr %) "ipc_UUID") @metadata))]
+                                   (delay (get uuid-meta :value))))]
+           (when-let [metadata (if cache?
+                                 (jargon/cached-get-metadata irods path)
+                                 (jargon/get-metadata irods path :known-type @(object-type irods user zone path)))]
+             (or
+               (get-from-metadata metadata)
+               nil))))
+     path]))
+
+(defn permission
+  [irods user zone path]
+  (cached-or-get irods
+    [from-listing (comp jargon-perms/fmt-perm :access_type_id) user zone path]
+    [(fn [cache? irods user path]
+       (if cache?
+         (jargon/cached-permission-for irods user path)
+         (delay
+           @(jargon/permission-for irods user path :known-type @(object-type irods user zone path)))))
```

Comment:
`:known-type` is one of the un-pushed improvements (used above as well). Many clj-jargon functions operate as "find the object type (collection or data object), then call the appropriate function for that type". We can save some work if we just tell the functions what type to use. Here, we use `object-type` because that can a.) use the ICAT if available, not that it's likely to use the ICAT for that but jargon for this, but more importantly b.) can use a cached value.